<a href="https://colab.research.google.com/github/Varshith-Tummala/Simulation-and-modeling-CS6647XTIA/blob/main/homework5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The states

In [42]:
Activity_state = {
    0: 'rest',
    1: 'sweeties',
    2: 'exercises'
}
state

{0: 'rest', 1: 'sweeties', 2: 'exercises'}

The transition matrix ( for the given state diagram) is as follows:

In [43]:
transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])
print(transition_matrix)

[[0.2 0.2 0.6]
 [0.2 0.1 0.7]
 [0.1 0.3 0.6]]


The random walk can be:

In [44]:
# number of steps and starting state
n = 15
starting_state = 0
previous_state = starting_state

# Performing the random walk
for i in range(n):
    current_state = np.random.choice([0, 1, 2], p=transition_matrix[previous_state])
    print(Activity_state[current_state], "---->", end="")
    # Update the previous state
    previous_state = current_state


exercises ---->exercises ---->sweeties ---->exercises ---->exercises ---->sweeties ---->exercises ---->exercises ---->exercises ---->exercises ---->exercises ---->sweeties ---->sweeties ---->exercises ---->exercises ---->

1. The starting state is Rest. Calculate the probability of possible activity on the 15th day.
2. What about the state and probability after 10000 days?


In [45]:
import numpy as np

# Transition matrix
transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])
# Starting state vector
start_state = np.array([1, 0, 0])  # Rest is the starting state
prob_15_days = np.linalg.matrix_power(transition_matrix.T, 15) @ start_state # Calculating probability after 15 days
prob_10000_days = np.linalg.matrix_power(transition_matrix.T, 10000) @ start_state # Calculating probability after 10,000 days
print("Probability of activities on the 15th day:") # Printing the respective probabilities
print("Rest:", prob_15_days[0])
print("Sweeties:", prob_15_days[1])
print("Exercises:", prob_15_days[2])

print("\nProbability of activities after 10,000 days:")
print("Rest:", prob_10000_days[0])
print("Sweeties:", prob_10000_days[1])
print("Exercises:", prob_10000_days[2])


Probability of activities on the 15th day:
Rest: 0.13761467889913792
Sweeties: 0.23853211009159786
Exercises: 0.6238532110092637

Probability of activities after 10,000 days:
Rest: 0.1376146788990357
Sweeties: 0.23853211009166178
Exercises: 0.6238532110089616


3. What do you observe from the above two?

**ANSWER:** when observed, we can see that as the number of days goes up, the odds tend to stay the same. After a long time, in this case 10,000 days, the chances of being in each state get close to certain values. In this case, the chances after 10,000 days are very close to the chances after 15 days. This shows that the distribution has reached a stable state. As the number of steps goes up, the values of probability states get closer together. If the number of trials is low, the model tends to give too high of odds for rest and sweets. As the number of trials goes up, we can see that they settle on 13.76% for rest, 23.85% for sweets, and 62.39% for exercises.